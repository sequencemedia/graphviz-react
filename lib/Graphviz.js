import React, { createRef, useEffect, useCallback } from 'react';
import classnames from 'classnames';
import { graphviz } from 'd3-graphviz';
const defaultOptions = {
    useWorker: false
};
export default function Graphviz({ dot, className, options = {}, onClick: handleClick = () => { } }) {
    const ref = createRef();
    useEffect(() => {
        const { current } = ref;
        if (current !== null) {
            graphviz(current, {
                ...defaultOptions,
                ...options
            }).renderDot(dot);
        }
    }, [dot, options]);
    const onClick = useCallback((event) => {
        const { current } = ref;
        if (current !== null) {
            const { target } = event;
            if (target instanceof Element) {
                if (current.contains(target)) {
                    handleClick(event);
                }
            }
        }
    }, [dot, options]);
    return (React.createElement("div", { className: classnames('graphviz', className), onClick: onClick, ref: ref }));
}
