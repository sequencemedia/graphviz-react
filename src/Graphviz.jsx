import React, {
  createRef,
  useEffect,
  useCallback
} from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {
  graphviz
} from 'd3-graphviz'

const defaultOptions = {
  useWorker: false
}

export default function Graphviz ({ graphRef: ref, dot, className, options, onClick: handleClick }) {
  useEffect(() => {
    if (ref) {
      const {
        current
      } = ref

      if (current) {
        graphviz(current, {
          ...defaultOptions,
          ...options
        }).renderDot(dot)
      }
    }
  }, [dot, options])

  const onClick = useCallback((event) => {
    const {
      current
    } = ref

    if (current) {
      const {
        target
      } = event

      if (current.contains(target)) {
        handleClick(event)
      }
    }
  }, [dot, options])

  return (
    <div
      className={classnames('graphviz', className)}
      onClick={onClick}
      ref={ref}
    />
  )
}

Graphviz.propTypes = {
  graphRef: PropTypes.shape({
    current: PropTypes.shape()
  }),
  dot: PropTypes.string.isRequired,
  className: PropTypes.string,
  options: PropTypes.shape(),
  onClick: PropTypes.func
}

Graphviz.defaultProps = {
  graphRef: createRef(),
  options: {},
  onClick () {
    //
  }
}
