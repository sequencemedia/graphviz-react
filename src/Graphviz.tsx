import React, {
  createRef,
  useEffect,
  useCallback
} from 'react'
import classnames from 'classnames'
import type { GraphvizOptions } from 'd3-graphviz'
import { graphviz } from 'd3-graphviz'

export interface IGraphvizProps {
  /**
   *  A React `ref` for the container
   */
  graphRef: React.RefObject<HTMLDivElement>
  /**
   * A string containing a graph representation using the Graphviz DOT language.
   * @see https://graphviz.org/doc/info/lang.html
   */
  dot: string
  /**
   * Options to pass to the Graphviz renderer.
   */
  options?: GraphvizOptions
  /**
   * The classname to attach to this component for styling purposes.
   */
  className?: string
  /**
   *  A handler for click events
   */
  onClick?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
}

const defaultOptions: GraphvizOptions = {
  useWorker: false
}

export default function Graphviz ({ graphRef: ref = createRef<HTMLDivElement>(), dot, className, options = {}, onClick: handleClick = () => {} }: IGraphvizProps): JSX.Element {
  useEffect(() => {
    const {
      current
    } = ref

    if (current !== null) {
      graphviz(current, {
        ...defaultOptions,
        ...options
      }).renderDot(dot)
    }
  }, [dot, options])

  const onClick = useCallback((event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    const {
      current
    } = ref

    if (current !== null) {
      const {
        target
      } = event

      if (target instanceof Element) {
        if (current.contains(target)) {
          handleClick(event)
        }
      }
    }
  }, [dot, options, handleClick])

  return (
    <div
      className={classnames('graphviz', className)}
      onClick={onClick}
      ref={ref}
    />
  )
}
