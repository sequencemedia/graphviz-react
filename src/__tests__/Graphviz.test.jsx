import React from 'react'
import * as renderer from 'react-test-renderer'
import Graphviz from '../Graphviz.jsx'

describe('<Graphviz />', () => {
  it('matches the snapshot', () => {
    const graphviz = renderer.create(
      <Graphviz
        dot='graph { a -- b }'
      />
    )

    expect(graphviz.toJSON())
      .toMatchSnapshot()
  })
})
