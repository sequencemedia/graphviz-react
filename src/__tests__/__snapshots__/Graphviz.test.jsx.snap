// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Graphviz /> matches the snapshot 1`] = `
<div
  className="graphviz"
  onClick={[Function]}
/>
`;
